<a href="https://colab.research.google.com/github/Sreenathkk00/DataScience-Projects/blob/main/House_Price.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [120]:
# Import Libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_percentage_error

In [121]:
# Read Dataset
house = pd.read_csv('https://github.com/YBI-Foundation/Dataset/raw/main/House%20Prices.csv')

In [122]:
# View the head of the dataset
house.head()

ID             Date     Price  Bedrooms  Bathrooms  Sqft_living  Sqft_lot  \
0   1  20140916T000000  280000.0         6       3.00         2400      9373   
1   2  20150422T000000  300000.0         6       3.00         2400      9373   
2   3  20140508T000000  647500.0         4       1.75         2060     26036   
3   4  20140811T000000  400000.0         3       1.00         1460     43000   
4   5  20150401T000000  235000.0         3       1.00         1430      7599   

   Floors  Waterfront  View  ...  Grade  Sqft_above  Sqft_basement  Yr_built  \
0     2.0           0     0  ...      7        2400              0      1991   
1     2.0           0     0  ...      7        2400              0      1991   
2     1.0           0     0  ...      8        1160            900      1947   
3     1.0           0     0  ...      7        1460              0      1952   
4     1.5           0     0  ...      6        1010            420      1930   

   Yr_renovated  zipcode      Lat     Long  Sqft_living15  Sqft_lot15  
0             0    98002  47.3262 -122.214           2060        7316  
1             0    98002  47.3262 -122.214           2060        7316  
2             0    98166  47.4444 -122.351           2590       21891  
3             0    98166  47.4434 -122.347           2250       20023  
4             0    98168  47.4783 -122.265           1290       10320  

[5 rows x 21 columns]

In [123]:
# Dataset Information
house.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ID             21613 non-null  int64  
 1   Date           21613 non-null  object 
 2   Price          21613 non-null  float64
 3   Bedrooms       21613 non-null  int64  
 4   Bathrooms      21613 non-null  float64
 5   Sqft_living    21613 non-null  int64  
 6   Sqft_lot       21613 non-null  int64  
 7   Floors         21613 non-null  float64
 8   Waterfront     21613 non-null  int64  
 9   View           21613 non-null  int64  
 10  Condition      21613 non-null  int64  
 11  Grade          21613 non-null  int64  
 12  Sqft_above     21613 non-null  int64  
 13  Sqft_basement  21613 non-null  int64  
 14  Yr_built       21613 non-null  int64  
 15  Yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  Lat            21613 non-null  float64
 18  Long  

In [124]:
#clean the dataset
house = house.drop(columns=['ID', 'Date'])



In [125]:
# Describe the dataset
house.describe()

Price      Bedrooms     Bathrooms   Sqft_living      Sqft_lot  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  2.161300e+04   
mean   5.401822e+05      3.370842      2.114757   2079.899736  1.510697e+04   
std    3.673622e+05      0.930062      0.770163    918.440897  4.142051e+04   
min    7.500000e+04      0.000000      0.000000    290.000000  5.200000e+02   
25%    3.219500e+05      3.000000      1.750000   1427.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068800e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             Floors    Waterfront          View     Condition         Grade  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean       1.494309      0.007542      0.234303      3.409430      7.656873   
std        0.539989      0.086517      0.766318      0.650743      1.175459   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         Sqft_above  Sqft_basement      Yr_built  Yr_renovated       zipcode  \
count  21613.000000   21613.000000  21613.000000  21613.000000  21613.000000   
mean    1788.390691     291.509045   1971.005136     84.402258  98077.939805   
std      828.090978     442.575043     29.373411    401.679240     53.505026   
min      290.000000       0.000000   1900.000000      0.000000  98001.000000   
25%     1190.000000       0.000000   1951.000000      0.000000  98033.000000   
50%     1560.000000       0.000000   1975.000000      0.000000  98065.000000   
75%     2210.000000     560.000000   1997.000000      0.000000  98118.000000   
max     9410.000000    4820.000000   2015.000000   2015.000000  98199.000000   

                Lat          Long  Sqft_living15     Sqft_lot15  
count  21613.000000  21613.000000   21613.000000   21613.000000  
mean      47.560053   -122.213896    1986.552492   12768.455652  
std        0.138564      0.140828     685.391304   27304.179631  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471000   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.230000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

In [126]:
# Columns
house.columns

Index(['Price', 'Bedrooms', 'Bathrooms', 'Sqft_living', 'Sqft_lot', 'Floors',
       'Waterfront', 'View', 'Condition', 'Grade', 'Sqft_above',
       'Sqft_basement', 'Yr_built', 'Yr_renovated', 'zipcode', 'Lat', 'Long',
       'Sqft_living15', 'Sqft_lot15'],
      dtype='object')

In [127]:
# Define target (y) and feature (x)
y = house['Price']  # Target variable
x = house.drop('Price', axis=1)  # Features


In [128]:
# Check shapes of x and y
x.shape,y.shape

((21613, 18), (21613,))

In [129]:
# Split the data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=2529)

# Check the shapes of the training and testing sets
print(f"Shape of x_train: {x_train.shape}")
print(f"Shape of x_test: {x_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of x_train: (15129, 18)
Shape of x_test: (6484, 18)
Shape of y_train: (15129,)
Shape of y_test: (6484,)


# **Predicting**

# LinearRegression

In [130]:

# Create a LinearRegression object
model = LinearRegression()


In [131]:
# Fit to the model train data
model.fit(x_train, y_train)



LinearRegression()

In [132]:
# Find the intercept value of the regression line
model.intercept_

6717685.627376461

In [133]:
# Find the slope
model.coef_


array([-3.10596119e+04,  3.52413772e+04,  1.09152073e+02,  1.37381856e-01,
        1.15075646e+04,  6.49425669e+05,  5.20277629e+04,  2.59941576e+04,
        9.61916002e+04,  7.03293463e+01,  3.88227268e+01, -2.59877157e+03,
        2.12577473e+01, -5.83034776e+02,  5.96977517e+05, -2.16839301e+05,
        2.03801644e+01, -3.68460864e-01])

In [134]:
# Using test data for prediction
y_pred = model.predict(x_test)
y_pred

array([ 214451.92155914,   53641.79747694,  366075.52128041, ...,
       1128507.75128272,  396731.29714253,  470850.17064903])

In [135]:
# Accuracy - Mean Absolute Percentage Error (MAPE)

error = mean_absolute_percentage_error(y_test, y_pred)
error


0.25602269508634157

In [136]:
# Calculate accuracy
accuracy = 100 - error * 100
accuracy

74.39773049136585

# Randon Forest Regression

In [137]:
# Initialize RandomForestRegressor Model
model = RandomForestRegressor(n_estimators=100, random_state=2529)

In [138]:
# Fit the model
model.fit(x_train, y_train)

RandomForestRegressor(random_state=2529)

In [139]:
# Prediction using x_test
y_pred = model.predict(x_test)
y_pred

array([ 344880.7       ,  213747.15666667,  404733.5       , ...,
       1483138.33333333,  500424.74      ,  589681.        ])

In [144]:
# Calculate Mean Absolute Percentage Error (MAPE)
error = mean_absolute_percentage_error(y_test, y_pred)
error

0.12955697517401707

In [142]:
#Accuracy
accuracy = 100 - error * 100
accuracy

87.0443024825983

**Linear Regression**

*  Mean Absolute Percentage Error (MAPE): 0.2560

* Accuracy: 74.39%



**Random Forest Regression**



*   Mean Absolute Percentage Error (MAPE): 0.1296
*   Accuracy: 87.04%

**Conclusion**

The Random Forest Regression model has the highest accuracy of 87.04% among the tested models. This model seems to capture the most variance in house prices and should be preferred for predicting house prices.




